contents/blog/feature-flags-as-a-service.md

```diff
@@ -0,0 +1,126 @@
+date: 2023-09-01
+author: ["ian-vanagas"]
+showTitle: true
+rootpage: /blog
+sidebar: Blog
+hideAnchor: true
+featuredImage: ../images/blog/green-blog-image.jpg
+featuredImageType: full
+category: Engineering
+tags:
+ - Feature flags
+ - Explainers
+---
+
+Feature flags are a tool to conditionally turn on or off code and components. They enable developers to safely roll out and roll back new features. This helps them follow the [best practice of disconnecting deployment from release](/blog/github-gitlab-feature-flags#why-do-github-and-gitlab-use-feature-flags).
+
+The basic functionality of feature flags is simple enough to build yourself. The challenge comes as you scale and the features you need become more complex. This post goes over the functionality of a feature flag service, the challenges of using a house-built service, and the benefits of feature flags as a service.
+
+## What does a feature flag service do?
+
+At their most simple, a feature flag service is:
+
+- A store of feature flags with keys with associated values.
+- A function that takes a key, checks the store, and returns a value.
+
+In Python, this might look like:
+
+```python
+class FeatureFlagService:
+    def __init__(self):
+        self.feature_flags = {
+            "new_feature": True
+        }
+
+    def is_enabled(self, feature_name):
+        return self.feature_flags.get(feature_name, False)
+
+if __name__ == "__main__":
+    feature_flag_service = FeatureFlagService()
+
+    if feature_flag_service.is_enabled("new_feature"):
+        print("New feature enabled")
+```
+
+This enables you to simply toggle on or off features from a centralized location.
+
+## The challenge of rolling your own feature flag service
+
+The big problem with this simple service is that the values are hard coded and require a redeploy to change. It isn’t remotely configurable, meaning it can’t change at runtime and doesn’t provide many benefits over using `if` statements where needed.
+
+The path to solving this is an increasingly complicated one, each step with its downsides:
+
+1. Use a config to store the flag values. This still requires a redeploy to change. Read more in [Feature flags vs configuration: Which should you choose?](/blog/feature-flags-vs-configuration)
+2. Store flag values in your database. This requires an admin panel or root access to modify the values and puts stress on your infrastructure. Database issues also impact the resiliency of flags.
+3. Write an external service or use a library to manage and evaluate feature flags. New dependencies creates a bunch of maintenance and optimization work to ensure they remain fast and bug-free. 
+
+As is apparent, rolling your own in-house feature flag service becomes increasingly complicated. We haven’t even touched on adding logic for targeting, caching for speed, or resiliency for parts of the service going down. All of these make you more likely to introduce bugs and create tech debt. 
```

Comment:
I think the first sentence is a bit repetitive of the opening paragraph of this section
```suggestion
We haven’t even touched on adding logic for targeting, caching for speed, or resiliency for parts of the service going down. All of these make you more likely to introduce bugs and create tech debt. 
```